src/transaction/log_replication.cpp

```diff
@@ -422,16 +444,96 @@ namespace cublog
       }
   }
 
-  template <typename T>
   void replicator::register_assigned_mvccid (TRANID tranid)
   {
     assert (m_replicate_mvcc);
 
-    const LOG_REC_ASSIGNED_MVCCID log_rec = m_redo_context.m_reader.reinterpret_copy_and_add_align<T> ();
+    m_redo_context.m_reader.advance_when_does_not_fit (sizeof (LOG_REC_ASSIGNED_MVCCID));
+    const LOG_REC_ASSIGNED_MVCCID log_rec =
+	    m_redo_context.m_reader.reinterpret_copy_and_add_align<LOG_REC_ASSIGNED_MVCCID> ();
 
     m_replicator_mvccid->new_assigned_mvccid (tranid, log_rec.mvccid);
   }
 
+  void replicator::replicate_sysop_end (TRANID tranid, const log_lsa &rec_lsa, const LOG_REC_SYSOP_END &log_rec)
+  {
+    assert (m_replicate_mvcc);
+
+    LOG_SYSOP_END_TYPE_CHECK (log_rec.type);
+    // TODO: according to implementation in log_sysop_commit_internal, these asserts should hold
+    // for other than sysop commit
+    //assert (!LSA_ISNULL (&log_rec.lastparent_lsa));
+    //assert (LSA_LT (&log_rec.lastparent_lsa, &rec_lsa));
+
+    if (log_rec.type == LOG_SYSOP_END_LOGICAL_MVCC_UNDO)
+      {
+	assert (!LSA_ISNULL (&log_rec.lastparent_lsa));
+	assert (LSA_LT (&log_rec.lastparent_lsa, &rec_lsa));
+
+	// mvccid might be valid or not
+	if (MVCCID_IS_VALID (log_rec.mvcc_undo_info.mvcc_undo.mvccid))
+	  {
+	    if (prm_get_bool_value (PRM_ID_ER_LOG_PTS_REPL_DEBUG))
+	      {
+		_er_log_debug (ARG_FILE_LINE, "[REPLICATOR_MVCC] %s tranid=%d MVCCID=%llu parent_MVCCID=%llu\n",
+			       log_sysop_end_type_string (log_rec.type), (int)tranid,
+			       (unsigned long long)log_rec.mvcc_undo_info.mvcc_undo.mvccid,
+			       (unsigned long long)log_rec.mvcc_undo_info.parent_mvccid);
+	      }
+	    m_replicator_mvccid->new_assigned_mvccid (tranid, log_rec.mvcc_undo_info.mvcc_undo.mvccid,
+		log_rec.mvcc_undo_info.parent_mvccid);
+	  }
+      }
+    else if (log_rec.type == LOG_SYSOP_END_COMMIT)
+      {
+	if (prm_get_bool_value (PRM_ID_ER_LOG_PTS_REPL_DEBUG))
+	  {
+	    _er_log_debug (ARG_FILE_LINE, "[REPLICATOR_MVCC] %s tranid=%d\n",
+			   log_sysop_end_type_string (log_rec.type), tranid);
+	  }
+	// only complete sub-ids, if found
+	m_replicator_mvccid->complete_sub_mvcc (tranid);
+      }
+    else if (log_rec.type == LOG_SYSOP_END_ABORT)
+      {
+	if (prm_get_bool_value (PRM_ID_ER_LOG_PTS_REPL_DEBUG))
+	  {
+	    _er_log_debug (ARG_FILE_LINE, "[REPLICATOR_MVCC] %s tranid=%d\n",
+			   log_sysop_end_type_string (log_rec.type), tranid);
+	  }
+	// only complete sub-ids, if found
+	m_replicator_mvccid->complete_sub_mvcc (tranid);
+      }
+    else
+      {
+	// nothing
```

Comment:
Is an assert possible here ?

Reply:
I think the `else` can be hit; just that there is nothing to be done. The else and logging is temporary anyway.